test/servinge2e/tracing_test.go

```diff
@@ -0,0 +1,236 @@
+		return fmt.Errorf("error dialing grpc to 127.0.0.1:%d: %w", portForward.LocalPort, err)
+	}
+	defer conn.Close()
+
+	queryClient := jaegerapi.NewQueryServiceClient(conn)
+
+	// First list all services so that we can find the one matching the prefix we know (as we only know the ksvc name).
+	getServicesResponse, err := queryClient.GetServices(context.Background(), &jaegerapi.GetServicesRequest{})
+	if err != nil {
+		return fmt.Errorf("error getting services: %w", err)
+	}
+
+	var serviceName string
+	for _, service := range getServicesResponse.Services {
+		ctx.T.Logf("service: %s", service)
+		if strings.HasPrefix(service, traceServiceNamePrefix) {
+			if serviceName != "" {
+				return fmt.Errorf("service: %s: found more than one service with %q prefix in Jaeger", service, traceServiceNamePrefix)
+			}
+			serviceName = service
+		}
+	}
+
+	if serviceName == "" {
+		return fmt.Errorf("didn't find any services with %q prefix in Jaeger", traceServiceNamePrefix)
+	}
+
+	// Find all traces matching our service name and the given operation name.
+	traceClient, err := queryClient.FindTraces(context.Background(), &jaegerapi.FindTracesRequest{
+		Query: &jaegerapi.TraceQueryParameters{
+			OperationName: traceOperationName,
+			ServiceName:   serviceName,
+		},
+	})
+	if err != nil {
+		return fmt.Errorf("error getting FindTraces client: %w", err)
+	}
+
+	traces := make(map[string][]jaegermodel.Span)
+
+	// Spans from a single trace can be in different chunks, so gather all traces together first.
+	for {
+		chunk, err := traceClient.Recv()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("error recv traces: %w", err)
+		}
+
+		for _, span := range chunk.Spans {
+			traces[span.TraceID.String()] = append(traces[span.TraceID.String()], span)
+		}
+	}
+
+	// Count all the matched traces.
+	count := 0
+	for traceID, spans := range traces {
+		count++
+
+		// All the serviceNamePrefixes should be covered by some traces, so we'll note the matched ones in a boolean array.
+		found := make([]bool, len(serviceNamePrefixes))
+
+		ctx.T.Logf("Trace %s:", traceID)
+
+		for _, span := range spans {
+			ctx.T.Logf("  %s(%s)", span.Process.ServiceName, span.OperationName)
+
+			matchesAny := false
+			for i, serviceNamePrefix := range serviceNamePrefixes {
+				if strings.HasPrefix(span.Process.ServiceName, serviceNamePrefix) {
+					matchesAny = true
+					found[i] = true
+				}
+			}
+
+			// Verify there is no span in the traces that doesn't match any of the serviceNamePrefixes.
+			if !matchesAny {
+				return fmt.Errorf("span %s(%s) doesn't match any of the expected prefixes (%v)", span.Process.ServiceName, span.OperationName, serviceNamePrefixes)
+			}
+		}
+
+		// Verify trace spans cover all services in serviceNamePrefixes.
+		for i, serviceNamePrefix := range serviceNamePrefixes {
+			if !found[i] {
+				return fmt.Errorf("Trace does not contain a span matching serviceName prefix %q", serviceNamePrefix)
+			}
+		}
```

Comment:
I feel these nested `for` blocks a little bit deeper. Can we split into some functions? I also want to the code what we are verifying by the function name.